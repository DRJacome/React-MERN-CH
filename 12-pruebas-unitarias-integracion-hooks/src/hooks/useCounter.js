import { useState } from "react"

export const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState)

  const reset = () => {
    setCounter(initialState)
  }

  const incrementar = () => {
    setCounter(counter + 1)
  }
  const decrementar = () => {
    setCounter(counter - 1)
  }

  return {
    state: counter,
    incrementar,
    decrementar,
    reset
  }
}