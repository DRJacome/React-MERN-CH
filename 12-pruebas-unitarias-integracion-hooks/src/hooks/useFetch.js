import { useEffect, useRef, useState } from "react"

export const useFetch = (url) => {
  const isMounted = useRef(true)
  const [estado, setEstado] = useState({ data: null, loading: true, error: null });

  useEffect(() => {
    return () => {
      isMounted.current = false
    }
  }, [])

  useEffect(() => {
    setEstado({ data: null, loading: true, error: null })
    fetch(url)
      .then(respuesta => respuesta.json())
      .then(data => {
        if (isMounted) {
          setTimeout(() => {
            setEstado({
              loading: false,
              error: null,
              data
            });
          }, 100);
        } else {
          console.log('setState() no se llamó');
        }
      })
      .catch(error => {
        setEstado({
          data: null,
          loading: false,
          error: "No se pudo obtener la información"
        })
      });
  }, [url])

  return estado;
}
