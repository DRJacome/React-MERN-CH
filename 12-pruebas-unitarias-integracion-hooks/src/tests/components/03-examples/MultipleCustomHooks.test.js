import React from 'react'
import { shallow } from 'enzyme'
import { MultipleCustomHooks } from "../../../components/03-ejemplos/MultipleCustomHooks"
import { useFetch } from '../../../hooks/useFetch'
import { useCounter } from '../../../hooks/useCounter'

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('Pruebas en <MultipleCustomHooks/>', () => {

  useCounter.mockReturnValue({
    counter: 10,
    incrementar: () => { }
  })

  test('Debe mostrarse correctamente', () => {

    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })
    const wrapper = shallow(<MultipleCustomHooks />)
    expect(wrapper).toMatchSnapshot();
  })

  test('Debe mostrar la informaciÃ³n', () => {

    useFetch.mockReturnValue({
      data: {
        activity: 'Read a formal research paper on an interesting subject',
        type: 'Education'
      },
      loading: false,
      error: null
    })

    const wrapper = shallow(<MultipleCustomHooks />)

    expect(wrapper.find('alert').exists()).toBe(false);
    expect(wrapper.find('.mb-0').text().trim()).toBe('Read a formal research paper on an interesting subject');

    expect(wrapper.find('footer').text().trim()).toBe('Education');
    console.log(wrapper.html());
  })
})