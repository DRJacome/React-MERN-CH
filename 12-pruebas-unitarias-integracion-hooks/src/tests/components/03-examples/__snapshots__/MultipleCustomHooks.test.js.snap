// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks/> Debe mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    Consejos The Bored API
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Cambiar frase
  </button>
</Fragment>
`;
